torch/onnx/symbolic_opset11.py

```diff
@@ -1558,3 +1560,45 @@ def prim_constant_chunk(g: jit_utils.GraphContext, self, chunks, dim):
         res.append(g.op("Slice", self, start, end, axis))
         start = end
     return res
+
+
+@_onnx_symbolic("aten::hstack")
+@_beartype.beartype
+def hstack(g: jit_utils.GraphContext, tensor_list: _C.Value):
+    # TODO: Doesn't support aten::at_least1d
+
+    first_tensor = g.op(
+        "SequenceAt",
+        tensor_list,
+        g.op("Constant", value_t=torch.tensor(0, dtype=torch.long)),
+    )
+    first_tensor_shape = g.op("Shape", first_tensor)
+    first_tensor_dim = g.op("Size", first_tensor_shape)
+
+    const_one = g.op("Constant", value_t=torch.tensor(1, dtype=torch.long))
+    equal_to_one = g.op("Equal", first_tensor_dim, const_one)
+
+    (
+        if_op_greater,
+        (if_context_equal, else_context_equal),
+        _,
+    ) = jit_utils.add_op_with_blocks(g, "If", equal_to_one, n_blocks=2, outputs=1)
+    result_if = if_context_equal.op(
+        "ConcatFromSequence", tensor_list, axis_i=0, new_axis_i=0
+    )
+    utils._add_output_to_block(if_context_equal.block, result_if)
+    result_else = else_context_equal.op(
+        "ConcatFromSequence", tensor_list, axis_i=1, new_axis_i=0
+    )
+    utils._add_output_to_block(else_context_equal.block, result_else)
+    result = if_op_greater.node().output()
+
+    return result
+
+
+@_onnx_symbolic("aten::vstack")
+@_beartype.beartype
+def vstack(g: jit_utils.GraphContext, tensor_list: _C.Value):
```

Comment:
For symbolic functions, an old pattern is to look if the tensor list is a static list, i.e., prim::ListConstruct. And if so, individual checks and validation similar to `at_least2d` can be applied. This usually results in a more efficient graph, but we could also argue it applies constant folding/optimization prematurely.

Taking a step back, it seems logical to support `aten::at_least2d`, and invoke it here. I feel this is also the best solution for torchlib, plus post optimizations. For that we optionally need OnnxFunction calling OnnxFunction in torchlib, and need ability to write loop over Sequence in OnnxFunction.

Reply:
I thought about unpack the list to for loop the tensors. But another difficulty is that a IF condition has to be set for each tensor to see if they are at least 2D or not (basically the whole at_least2D). If this part of logic is also written in constant, Then 2D promotion would be added into graph regardless of next inputs are 2D already.

Reply:
yea, usually it is accompanied with the assumption that input ranks remain the same. Hence more suited for post optimization.

Reply:
OK I understand it now. It's kind of like the idea of op.Transpose that we usually flip them the same way in all batches.

Reply:
Done